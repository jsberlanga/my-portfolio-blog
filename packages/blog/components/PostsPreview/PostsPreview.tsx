import * as React from 'react';
import Link from 'next/link';
import { css } from '@emotion/react';
import { GoNext } from '@juliosoto/components/Icons';
import { TPostPreview } from '@juliosoto/lib/types';
import { Timestamp } from '@juliosoto/components';
import dynamic from 'next/dynamic';
import { motion } from 'framer-motion';
import { variants } from '@juliosoto/lib/styles';

const DynamicVisits = dynamic(() => import('../Visits'));

const styles = css`
  max-width: var(--content-width);
  margin: 0 auto var(--gap);

  button {
    font-weight: 600;
  }
`;

interface PostPreviewProps {
  postsPreviewData: TPostPreview[];
}

const PostsPreview: React.FC<PostPreviewProps> = ({ postsPreviewData }) => {
  const [visits, setVisits] = React.useState(0);

  React.useEffect(() => {
    const getVisits = () =>
      postsPreviewData.map(async ({ slug }) => {
        try {
          const response = await fetch(`/api/visits/?slug=${slug}`, {
            method: 'GET',
          });
          if (response.status === 200) {
            const data = await response.json();

            setVisits(data.post.visits);
          }
        } catch (error) {
          console.log(error);
        }
      });
    getVisits();
  }, [postsPreviewData]);

  return (
    <motion.section
      variants={variants.fadeIn}
      initial="initial"
      animate="animate"
      css={styles}
    >
      {postsPreviewData.map((post, idx) => (
        <Link href={`/${post.slug}`} key={idx}>
          <a>
            <h4>{post.title}</h4>
            <Timestamp>Published on {post.publishedAt}</Timestamp>
            <DynamicVisits>{visits} views</DynamicVisits>
            <br />
            <p className="small">{post.summary}</p>
            <button>
              Read more <GoNext />
            </button>
          </a>
        </Link>
      ))}
    </motion.section>
  );
};

export default PostsPreview;
