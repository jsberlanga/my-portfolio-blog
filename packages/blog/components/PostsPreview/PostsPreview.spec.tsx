import * as React from 'react';
import { render } from '@testing-library/react';
import PostsPreview from './PostsPreview';

jest.mock('@juliosoto/components/Icons', () => ({
  GoNext: () => 'GoNext',
}));

const mockPostsData = [
  {
    title: 'test',
    summary: 'summary_test',
    slug: 'test',
    publishedAt: '2020-01-01',
    visits: 10,
  },
];

describe('PostsPreview', () => {
  test('renders correctly', () => {
    const { container } = render(
      <PostsPreview postsPreviewData={mockPostsData} />,
    );

    expect(container).toMatchSnapshot();
  });
});
