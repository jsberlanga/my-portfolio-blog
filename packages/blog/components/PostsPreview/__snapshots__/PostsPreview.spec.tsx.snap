// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsPreview renders correctly 1`] = `
.emotion-0 {
  max-width: var(--content-width);
  margin: 0 auto var(--gap);
}

.emotion-0 button {
  font-weight: 600;
}

.emotion-1 {
  color: var(--c-special);
  display: block;
}

<div>
  <section
    class="emotion-0"
    style="opacity: 0;"
  >
    <a
      href="/test"
    >
      <h4>
        test
      </h4>
      <time
        class="small emotion-1"
      >
        Published on 
        2020-01-01
      </time>
      <br />
      <p
        class="small"
      >
        summary_test
      </p>
      <button>
        Read more 
        GoNext
      </button>
    </a>
  </section>
</div>
`;
